### ISSUE TITLE
Disallow `Depends:`; require `Imports:` only.

### ISSUE THREAD

**Issue by mattdowle** [Rdatatable/data.table#3076]
I've only recently realized how bad `Depends:` is,  thanks to Jan's importing vignette.  I just made its discouragement stronger :  https://github.com/Rdatatable/data.table/pull/3069/commits/51590bcfec6e9cccc9271800bcf01749a42c7e80

We could disallow Depends. This would also be beneficial to `cedta()`'s awkward implementation on its last line where it needs to do the tryCatch just for packages which Depend; that line could be removed.

But before we disallow Depends, we'd need to ask 69 CRAN packages to change from Depends to Imports.  (Most revdeps already Import.)   The longer we leave it, the greater the potential for new packages using Depends to be added to CRAN and the harder it will be to change.

- [ ] Ac3net
- [ ] AF
- [ ] batchtools
- [ ] bea.R
- [ ] behavr
- [ ] BuyseTest
- [ ] cdparcoord
- [ ] cffdrs
- [ ] classifierplots
- [ ] clickstream
- [ ] corpustools
- [ ] cvAUC
- [ ] damr
- [ ] dfmeta
- [ ] drgee
- [ ] easycsv
- [ ] edgeRun
- [ ] EurosarcBayes
- [ ] gbp
- [ ] GenomicTools
- [ ] GenomicTools.fileHandler
- [ ] greport
- [ ] haploReconstruct
- [ ] heatwaveR
- [ ] heims
- [ ] ie2misc
- [ ] iemisc
- [ ] JWileymisc
- [ ] koRpus
- [ ] LabourMarketAreas
- [ ] lookupTable
- [ ] LSPFP
- [ ] miLineage
- [ ] mrMLM
- [ ] mrMLM.GUI
- [ ] multicastR
- [ ] musica
- [ ] networkR
- [ ] NNS
- [ ] openSTARS
- [ ] orgR
- [ ] panelaggregation
- [ ] partools
- [ ] penaltyLearning
- [ ] pkggraph
- [ ] QuantTools
- [ ] Rbitcoin        @jangorecki please could you change from Depends to Imports :-)
- [ ] reinsureR
- [ ] riskRegression
- [ ] Rnets
- [ ] robCompositions
- [ ] RSauceLabs
- [ ] RWildbook
- [ ] sentometrics
- [ ] simPop
- [ ] simstudy
- [ ] sitree
- [ ] skm
- [ ] slim
- [ ] sparseFLMM
- [ ] stranger
- [ ] surveyplanning
- [ ] tcpl
- [ ] ttwa
- [ ] twl
- [ ] validateRS
- [ ] vardpoor
- [ ] VIM
- [ ] word.alignment

**Comment by jangorecki** [comment]
It shouldn't be disallowed because there are valid use cases for Depends, for example my package [dtq](https://github.com/jangorecki/dtq) which is a logging mechanism for `[.data.table`. I recall other package too: dt.nuggets(?). We should definitely recommend to imports instead of depends, but IMO it should not be disallowed.

**Comment by mattdowle** [comment]
Fair enough. Why exactly is it that those packages need Depends and can't use Imports?

**Comment by jangorecki** [comment]
They are extensions of data.table, their existence without data.table is pointless. They not just "use" data.table.

**Comment by mattdowle** [comment]
I still don't follow. Importing is as strong as depending in that imports are required. No package importing data.table can install without it too. What would fail if they changed to Imports?

**Comment by jangorecki** [comment]
It is as strong, except the fact that imported namespace doesn't need to be attached. It doesn't mask, but it expects data.table to be attached. There are definitely very few use cases like this so general rule should be to import.

**Comment by mattdowle** [comment]
75 maintainers emailed today  (up from the 69 in Sep 2018) : 
```R
deps = tools::package_dependencies("data.table", which="Depends", reverse=TRUE)[[1]]
paste(sapply(deps, maintainer),collapse=";")
```

> Dear maintainer,
> 
> Hello! Thanks for using data.table. Your package is one of 75 CRAN packages that Depend on data.table. Another 469 Import data.table. Importing is much preferred.
> 
> The data.table importing vignette now contains this text :
> 
> "  Besides the Imports: field, you can also use Depends: data.table but we strongly discourage this approach (and may disallow it in future) because this loads data.table into your user’s workspace; i.e. it enables data.table functionality in your user’s scripts without them requesting that. Imports: is the proper way to use data.table within your package without inflicting data.table on your user. In fact, we hope the Depends: field is eventually deprecated in R since this is true for all packages. "
> 
> https://cran.r-project.org/web/packages/data.table/vignettes/datatable-importing.html
> 
> Another motivation is that it would be beneficial to `cedta()`'s awkward implementation on its last line where it needs to do the tryCatch just for packages which Depend. That line could be removed.
> 
> Eventually we'll add a message, then a warning, and after a few years possibly only allow Importing and not Depending on data.table.  The next time you update your package on CRAN, are you ok to change from Depends to Imports please?
> 
> We're tracking this issue here : https://github.com/Rdatatable/data.table/issues/3076
> 
> It's very early days on this issue and this is just a first email to let you know our thoughts and the plan.  Feedback and comments are very welcome, ideally directly in the GitHub issue.
> 
> Best, Matt
>

**Comment by mbannert** [comment]
I fully agree. I always use data.table for my backends and have imported on my more recent packages. In my case *panelaggregation* was a try to put a package on CRAN back when people wore pyjamas and lived life slow. Last time a data.table change caused issues I tried to get the package of CRAN but was discouraged to do so. I'd rather put my time into my newer packages than to fix a package which is more or less abandoned. On the other hand it would probably not be that much of deal to fix it though I haven't looked at the package in years. 

Bottom line: from my point of view, disallowing would be a bit too much since it would get my old package of CRAN forcefully.

**Comment by tdhock** [comment]
hi I just updated my penaltyLearning package, https://github.com/tdhock/penaltyLearning/commit/ef8e7fbdb1e240ad9dd7c8b40520cd9231ae0580. now it imports data.table instead of depends. basically I just had to add a bunch of library(data.table) lines in tests and examples.

**Comment by bozenne** [comment]
Hi thanks for the email and for data.table
I have also changed to imports in the Github version of the BuyseTest package.
Will be updated on CRAN at some point.

Best
brice

**Comment by MichaelChirico** [comment]
Adding these from Bioconductor:

```
- [ ] amplican
- [ ] Chicago
- [ ] chimeraviz
- [ ] GenoGAM
- [ ] GGtools
- [ ] GOTHiC
- [ ] metavizr
- [ ] OUTRIDER
- [ ] PGA
- [ ] QUALIFIER
- [ ] QuartPAC
- [ ] rBiopaxParser
- [ ] RCAS
- [ ] rfPred
- [ ] rTANDEM
- [ ] SNPhood
- [ ] TIN
```

Also added some code for repeating this & some more new packages from CRAN under `<details>` in OP

**Comment by brown-jason** [comment]
tcpl package has moved data.table to imports

**Comment by mattdowle** [comment]
Reminder email sent today to these 74 : 
```
 [1] "Ac3net"                   "AF"                       "batchtools"              
 [4] "bea.R"                    "behavr"                   "BuyseTest"               
 [7] "cdparcoord"               "classifierplots"          "clickstream"             
[10] "corpustools"              "CoSMoS"                   "cvAUC"                   
[13] "damr"                     "dfmeta"                   "drgee"                   
[16] "easycsv"                  "EBPRS"                    "edgeRun"                 
[19] "EurosarcBayes"            "gbp"                      "GenomicTools"            
[22] "GenomicTools.fileHandler" "GenoScan"                 "glmaag"                  
[25] "greport"                  "haploReconstruct"         "heims"                   
[28] "ie2misc"                  "iemisc"                   "JWileymisc"              
[31] "LabourMarketAreas"        "lookupTable"              "lori"                    
[34] "LSPFP"                    "miLineage"                "mrMLM"                   
[37] "mrMLM.GUI"                "multicastR"               "musica"                  
[40] "networkR"                 "openSTARS"                "orgR"                    
[43] "panelaggregation"         "partools"                 "phenocamapi"             
[46] "pkggraph"                 "PROSPER"                  "QuantTools"              
[49] "Rbitcoin"                 "rblt"                     "reinsureR"               
[52] "riskRegression"           "Rnets"                    "robCompositions"         
[55] "RSauceLabs"               "RWildbook"                "sentometrics"            
[58] "shinyML"                  "simPop"                   "sitreeE"                 
[61] "skm"                      "slim"                     "somspace"                
[64] "sparseFLMM"               "SVN"                      "synthACS"                
[67] "TreeLS"                   "ttwa"                     "Tushare"                 
[70] "twl"                      "validateRS"               "VIM"                     
[73] "WGScan"                   "word.alignment"  
```

---

> Dear maintainer,
> Hello! Thanks for using data.table. Your package is one of 74 CRAN packages that Depend on data.table. Another 667 Import data.table. Importing is much preferred.
> The data.table importing vignette now contains this text :" Besides the Imports: field, you can also use Depends: data.table but we strongly discourage this approach (and may disallow it in future) because this loads data.table into your user’s workspace; i.e. it enables data.table functionality in your user’s scripts without them requesting that. Imports: is the proper way to use data.table within your package without inflicting data.table on your user. In fact, we hope the Depends: field is eventually deprecated in R since this is true for all packages. "
https://cran.r-project.org/web/packages/data.table/vignettes/datatable-importing.html
Another motivation is that it would be beneficial to cedta()'s awkward implementation on its last line where it needs to do the tryCatch just for packages which Depend. That line could be removed.Eventually we'll add a message, then a warning, and after a few years possibly only allow Importing and not Depending on data.table. The next time you update your package on CRAN, are you ok to change from Depends to Imports please?
We're tracking this issue here : https://github.com/Rdatatable/data.table/issues/3076
It has been 6 months since I first emailed. Each email and package status is logged in the tracking issue. Feedback and comments are very welcome, ideally directly in the GitHub issue.
Best, Matt

**Comment by ekstroem** [comment]
Updated `networkR` package with Depends -> Imports on its way to CRAN

**Comment by jangorecki** [comment]
Daily refreshed:
[![Depends](https://jangorecki.gitlab.io/rdeps/data.table/Depends.svg?sanitize=true)](https://gitlab.com/jangorecki/rdeps) [![Imports](https://jangorecki.gitlab.io/rdeps/data.table/Imports.svg?sanitize=true)](https://gitlab.com/jangorecki/rdeps)
CRAN + BioC

**Comment by gyang274** [comment]
Hi, @mattdowle 

Thank you for the notice, and the wonderful data.table package.

I experienced inconsistent solution for communicating with Rcpp, when used imports instead of depends. This could caused by other issues. But at that time, depends is a reasonable choice as the package relied on and built around data.table features.

Let me test on imports and get back to you.

Thank you.

**Comment by sborms** [comment]
Thanks for the reminder! Should be done for the sentometrics package by next CRAN release.

Sam

**Comment by bozenne** [comment]
Hello, 
I have (finally) updated the version of the BuyseTest package on CRAN (1.8)
where data.table has been moved to the imports section, as you suggested.

Regards
brice

**Comment by JWiley** [comment]
Hello,

This is fixed in JWileymisc now propagating through CRAN. I took the opportunity to change all depends into imports.

Cheers,

Josh

**Comment by alexWhitworth** [comment]
`synthACS` now uses Imports instead of Depends. Push sent to CRAN.

**Comment by mattdowle** [comment]
Currently 79 packages that Depend rather than Import and they continue to be created or updated.

```
wget http://cloud.r-project.org/web/packages/packages.rds  # has more fields than PACKAGES.rds
x = readRDS("packages.rds")
rownames(x) = x[,"Package"]
deps = tools::package_dependencies("data.table", which="Depends", reverse=TRUE)[[1]]
length(deps)   # 79
y = as.matrix(table(substring(x[deps,"Published"],1,7)))
y = y[sort(rownames(y), decreasing=TRUE),,drop=FALSE]
y
# Number of packages Depending (not Importing) data.table by publish month
2022-10    5
2022-09    2
2022-08    3
2022-06    5
2022-05    2
2022-04    1
...
```

I never did get to the bottom of Jan's objections above to deprecating (very gently and slowly over several years) Depend'ing; i.e. what the valid use cases of Depend'ing are: why Import'ing won't work or is not as good as Depend'ing.

**Comment by MichaelChirico** [comment]
I used Gemini to file migration PRs against all public downstreams; I condensed that into a depends-to-imports SKILL: https://github.com/MichaelChirico/r-skills/blob/main/r-depends-to-imports/SKILL.md

I think that once 1.19.0 is released, we can begin to deprecate support for `Depends` downstreams in {data.table} to finally force this into practice.
